lib/wallet/src/Cardano/Wallet/Write/Tx.hs

```diff
@@ -0,0 +1,481 @@
+asAnyRecentEra = \case
+    Cardano.InAnyCardanoEra Cardano.BabbageEra a ->
+        Just $ InAnyRecentEra RecentEraBabbage a
+    Cardano.InAnyCardanoEra Cardano.AlonzoEra a ->
+        Just $ InAnyRecentEra RecentEraAlonzo a
+    _ -> Nothing
+
+--------------------------------------------------------------------------------
+-- TxIn
+--------------------------------------------------------------------------------
+
+type TxIn = Ledger.TxIn StandardCrypto
+
+-- | Useful for testing
+unsafeMkTxIn :: ByteString -> Word -> TxIn
+unsafeMkTxIn hash ix = Ledger.mkTxInPartial
+    (toTxId hash)
+    (fromIntegral ix)
+  where
+    toTxId :: ByteString -> Ledger.TxId StandardCrypto
+    toTxId h =
+        (Ledger.TxId (unsafeMakeSafeHash $ UnsafeHash $ toShort h))
+
+--------------------------------------------------------------------------------
+-- TxOut
+--------------------------------------------------------------------------------
+
+type TxOut era = Core.TxOut era
+
+type Address = Ledger.Addr StandardCrypto
+
+unsafeAddressFromBytes :: ByteString -> Address
+unsafeAddressFromBytes bytes = case Ledger.deserialiseAddr bytes of
+    Just addr -> addr
+    Nothing -> error "unsafeAddressFromBytes: failed to deserialise"
+
+scriptFromBytes
+    :: ByteString
+    -> Either CBOR.DecoderError (Script LatestLedgerEra)
+scriptFromBytes bs = CBOR.decodeAnnotator "Script" CBOR.fromCBOR (fromStrict bs)
+
+scriptFromCardanoScriptInAnyLang
+    :: Cardano.ScriptInAnyLang
+    -> (Script LatestLedgerEra)
+scriptFromCardanoScriptInAnyLang
+    = Cardano.toShelleyScript
+    . fromMaybe (error "all valid scripts should be valid in latest era")
+    . Cardano.toScriptInEra latestEra
+  where
+    latestEra = Cardano.BabbageEra
+
+scriptToCardanoScriptInAnyLang
+    :: Script LatestLedgerEra
+    -> Cardano.ScriptInAnyLang
+scriptToCardanoScriptInAnyLang =
+    rewrap
+    . Cardano.fromShelleyBasedScript latestEra
+  where
+    rewrap (Cardano.ScriptInEra _ s) = Cardano.toScriptInAnyLang s
+    latestEra = Cardano.ShelleyBasedEraBabbage
+
+scriptToCardanoEnvelopeJSON :: Script LatestLedgerEra -> Aeson.Value
+scriptToCardanoEnvelopeJSON = scriptToJSON . scriptToCardanoScriptInAnyLang
+  where
+    scriptToJSON
+        :: Cardano.ScriptInAnyLang
+        -> Aeson.Value
+    scriptToJSON (Cardano.ScriptInAnyLang l s) = Aeson.toJSON
+            $ obtainScriptLangConstraint l
+            $ Cardano.serialiseToTextEnvelope Nothing s
+      where
+        obtainScriptLangConstraint
+          :: Cardano.ScriptLanguage lang
+          -> (Cardano.IsScriptLanguage lang => a)
+          -> a
+        obtainScriptLangConstraint lang f = case lang of
+            (Cardano.SimpleScriptLanguage Cardano.SimpleScriptV1) -> f
+            (Cardano.SimpleScriptLanguage Cardano.SimpleScriptV2) -> f
+            (Cardano.PlutusScriptLanguage Cardano.PlutusScriptV1) -> f
+            (Cardano.PlutusScriptLanguage Cardano.PlutusScriptV2) -> f
+
+scriptFromCardanoEnvelopeJSON
```

Comment:
```suggestion
-- | Indirectly roundtrip-tested by "Api.TypesSpec"
scriptFromCardanoEnvelopeJSON
```

Reply:
Or, right, they aren't here, but they will be in #3436...

Reply:
I'm adding the missing tests... will push soon...

I discovered this gotcha though:
```
-- | NOTE: The type 'Script LatestLedgerEra' cannot distinguish between
-- 'Cardano.SimpleScriptV1' and 'Cardano.SimpleScriptV2'. This means the
-- roundtrip
-- @
--     scriptToCardanoScriptInAnyLang . scriptFromCardanoScriptInAnyLang
-- @
-- will convert 'SimpleScriptV1' to 'SimpleScriptV2'. Because 'SimpleScriptV1'
-- is 'ShelleyEra'-specific, and 'ShelleyEra' is not a 'RecentEra', this should
-- not be a problem.
```